Add help menu and PF About Modal

rebeccaalpert opened the pull request:
Modified Dropdown so we could have one without a caret, added Help dropdown to masthead, added PF About Modal to Help dropdown, and modified SoftwareDetails so it could be used in the modal.

I adjusted the header font sizes/spacing in the modal to be closer to the image; let me know if I should use default ones instead.

Fixes [CONSOLE-965](https://jira.coreos.com/browse/CONSOLE-965).

Dropdown:
<img width="1424" alt="screen shot 2018-10-31 at 10 03 34 am" src="https://user-images.githubusercontent.com/7014965/47794745-730c0300-dcf7-11e8-8387-70f591e1529e.png">

Modal:
<img width="1421" alt="screen shot 2018-10-31 at 10 02 42 am" src="https://user-images.githubusercontent.com/7014965/47794761-7a331100-dcf7-11e8-81f3-29c3ada95fa6.png">

Gif:
![dropdown](https://user-images.githubusercontent.com/7014965/47794770-815a1f00-dcf7-11e8-9a85-96c9ef0db0b2.gif)

@openshift/team-ux-review, could you please review? Thanks!

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+          {label: 'Documentation', callback: this.redirect},
+          {label: 'About', callback: this.open}]}
+        noButton
+        noCaret
```
Do we want to keep the caret? We had it in the previous console:

<img width="309" alt="screen shot 2018-10-31 at 11 46 01 am" src="https://user-images.githubusercontent.com/1167259/47800501-ca63a080-dd02-11e8-9ab7-95fde406aab6.png">

rebeccaalpert replied:
@serenamarie125 and @bmignano, do you want the caret?

cshinn replied:
[This doc](https://www.patternfly.org/pattern-library/application-framework/masthead/#design) says no caret

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+  }
+
+  redirect() {
+    window.open('https://docs.openshift.com/online/welcome/index.html', '_blank');
```
It would be much better to make this an actual link if we can. This has security implications because we can't set `rel=noreferrer`.

Note that the URL is configurable, so we can't hard-code it here. We need to use `openshiftHelpBase` from here:

https://github.com/openshift/console/blob/master/frontend/public/components/utils/documentation.tsx#L6-L7

rebeccaalpert replied:
The issue I was running into is that hrefs added to the Action Menu all turned into unusable links like this: http://localhost:9000/k8s/ns/myproject/https://docs.okd.io/latest/

I looked at the dropdown component and couldn't figure out why it was happening.

spadgett replied:
How were you building the link?

rebeccaalpert replied:
<ActionsMenu
        actions={[
          {label: 'Documentation', href: openshiftHelpBase},
          {label: 'About', callback: this.open}]}
        noButton
        noCaret
        title={<i className="fa fa-question-circle-o co-masthead__help-icon" />} />

spadgett replied:
Yeah, I think this will require changes to `ActionsMenu`. It's using `history.push`, which just adds to the path.

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+  }
+
+  _getUserInfo() {
+    coFetchJSON('api/kubernetes/apis/user.openshift.io/v1/users/~')
```
It would be better to do this once and keep it in Redux instead of multiple times for different nav components. Then the modal itself can connect to Redux to get the username rather than passing it down.

It might be simpler for now just to not include the username in the modal, however.

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+
+  render () {
+    const username = this.state.username;
+    return username ? <HelpMenu username={username} /> : null;
```
`username` won't always be defined, but we still want to show the help menu

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+    this.redirect = this.redirect.bind(this);
+  }
+
+  close() {
```
```suggestion
  closeAboutModal() {
```

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+    this.setState({ showModal: false });
+  }
+
+  open() {
```
```suggestion
  openAboutModal() {
```

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+export class HelpMenu extends React.Component<HelpMenuProps, HelpMenuState> {
+  constructor (props) {
+    super(props);
+    this.state = { showModal: false };
```
```suggestion
    this.state = { showAboutModal: false };
```

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+    this.setState({ showModal: true });
+  }
+
+  redirect() {
```
```suggestion
  openDocumentation() {
```

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+  }
+
+  if (props.flags[FLAGS.OPENSHIFT]) {
+    return <OSHelpMenu />;
```
We shouldn't need separate help menus here. The items should be the same regardless of whether it's OpenShift (at least for now).

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+        noButton
+        noCaret
+        title={<i className="fa fa-question-circle-o co-masthead__help-icon" />} />
+      <AboutModal className="co-masthead__modal" logo={okdLogoImg} altLogo="OKD" productTitle="OpenShift" show={this.state.showModal} onHide={this.close}>
```
We're not always showing the OKD logo. We need to show the right branding depending on what's running (OKD, OCP, Online, Dedicated)

spadgett commented on frontend/public/components/_masthead.scss:
```diff
@@ -36,13 +54,38 @@
     display: flex;
     height: 100%;
   }
+  &__modal {
```
I don't think we should need most of these styles as long as we're using the patternfly-react components to render the modal (`AboutModal.Versions`, etc)

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,96 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+        <h2>About</h2>
+        <p>OpenShift is Red Hat&apos;s container application platform that allows developers to quickly develop, host, and scale applications in a cloud environment.</p>
+        <h2>Version</h2>
+        <SoftwareDetails className="co-masthead__modal--details" kubernetesTitle="Kubernetes Master" noColumns openShiftFirst openShiftTitle="OpenShift Master" titleClassName="co-masthead__modal--subtitle" detailClassName="co-masthead__modal--version" />
```
Is this going to fire off the versions requests before you open the modal?

spadgett commented on frontend/public/components/software-details.jsx:
```diff
@@ -44,12 +44,12 @@ export const SubHeaderRow = ({header, children}) => {
-const SoftwareDetailRow = ({title, detail, text, children}) => {
-  return <div className="row cluster-overview-cell__info-row">
-    <div className="col-xs-6">
+const SoftwareDetailRow = ({className, title, detail, text, children, noColumns, titleClassName, detailClassName}) => {
```
I'm not sure we should reuse this component. We've talked about removing it, so I might just move the two requests to get the versions into the about modal.

spadgett commented on frontend/public/style/_config.scss:
```diff
@@ -10,3 +10,5 @@ $icon-font-path: $coreosWebFontPath + '/';
 // CoreOS-web image path.
 $img-path: $coreosWebDistPath + 'img';
+// Modal background image path.
+$img-path: '~patternfly/dist/img/'
```
Is this redefining the variable from just above?

rhamilto commented on frontend/public/components/_masthead.scss:
```diff
@@ -14,12 +14,30 @@
+    margin-left: auto;
+    margin-right: 20px;
+  }
+  &__help {
```
We will only want this to appear at desktop resolutions.  Should appear in the left nav for mobile.

rebeccaalpert replied:
It should be in the left nav with Workloads and everything?

rebeccaalpert replied:
Thanks, I'll hide it.

rhamilto commented on frontend/public/components/_masthead.scss:
```diff
@@ -14,12 +14,30 @@
+  }
+  &__help-icon {
+    font-size: 14px;
+    margin: 0px 2px 0px 5px;
```
Nit:  zero is always zero, so no need for `px` on `0px`.

rhamilto commented on frontend/public/vendor.scss:
```diff
@@ -49,6 +49,7 @@
 @import '~patternfly/dist/sass/patternfly/forms';
 @import '~patternfly/dist/sass/patternfly/dependencies/bootstrap-switch/bootstrap-switch';
 @import '~patternfly/dist/sass/patternfly/bootstrap-switch';
+@import '~patternfly/dist/sass/patternfly/_about-modal';
```
`_` is not necessary in the filename

rhamilto commented on frontend/public/components/_masthead.scss:
```diff
@@ -14,12 +14,30 @@
+    }
+  }
+  &__help-icon {
+    font-size: 14px;
```
The icon doesn't render great at this font size (looks a bit anemic).  
![screen shot 2018-10-31 at 3 25 04 pm](https://user-images.githubusercontent.com/895728/47813347-37862e80-dd21-11e8-9905-c70764703a2e.PNG)

rhamilto commented on frontend/public/components/_masthead.scss:
```diff
@@ -14,12 +14,30 @@
     height: $masthead-height-desktop;
   }
+  &__co-actions-menu {
+    padding: 0px 0px 0px 5px;
```
Nit: zero is always zero, so no need for `px` on `0px`.

serenamarie125 commented:
The current PF3 documentation has changed - @spadgett would you be ok with the following ?
- don't show a caret next to the help icon
- have a selected state for the help icon when menu is shown (this happens if you are using PF3 code from what i see)
- use a dropdown (no caret in the menu)

rebeccaalpert commented:
Addressed PR feedback.

Resized icon:
<img width="126" alt="screen shot 2018-10-31 at 10 46 20 pm" src="https://user-images.githubusercontent.com/7014965/47829644-dc732c80-dd5e-11e8-9f09-d95ded51f689.png">

Modal:
<img width="1396" alt="screen shot 2018-10-31 at 10 46 28 pm" src="https://user-images.githubusercontent.com/7014965/47829648-e137e080-dd5e-11e8-9c3f-218905e842bd.png">

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -9,12 +10,133 @@ import * as dedicatedLogoImg from '../imgs/openshift-dedicated-logo.svg';
+      }
+    }
+
+    _checkOpenShiftVersion() {
```
The version logic should be in the modal, not in the menu. We want to check the version when the modal is opened.

spadgett commented on frontend/public/style/_config.scss:
```diff
@@ -8,5 +8,5 @@ $fa-font-path: $coreosWebFontPath;
 // PatternFly font path.
 $font-path: '~patternfly/dist/fonts/';
-// CoreOS-web image path.
-$img-path: $coreosWebDistPath + 'img';
```
Before we remove this, do we know how this was used?

rebeccaalpert replied:
I talked to Jeff about how to do this and he said he didn't see it used anywhere.

serenamarie125 commented:
@rebeccaalpert the selected treatment doesn't match PF yet, and also the menu should align with the bottom of the masthead.  Take a look at the [PF React Example](https://rawgit.com/patternfly/patternfly-react/gh-pages/patternfly-3/index.html?knob-Nav%20Toggle=true&selectedKind=patternfly-react%2FApplication%20Framework%2FMasthead&selectedStory=Vertical%20Menu&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

rebeccaalpert commented:
Thanks for the link; I'll work on that.

bmignano commented:
There were some updates needed on the About modal from the original documentation – attaching Mary's updated mocks for the downstream version for reference.
![about-modal](https://user-images.githubusercontent.com/40238481/47931213-84ded900-dea4-11e8-81f3-3a7d53fd3cfc.png)

spadgett commented:
We probably want the title of the modal to match the branding.

rhamilto commented:
I thought the diagonal background was specific to RCUE?  If that's the case, we'd want to avoid using it in okd (the upstream).

beanh66 commented:
Thanks @rhamilto. Do you know what image is supposed to be there for upstream? If not, I'll defer to @serenamarie125 and @kybaker. Brie does have the downstream version of that screenshot with the RH logo as well, when needed.

rhamilto commented:
> Do you know what image is supposed to be there for upstream?

I assume it is the PatternFly default?
![screen shot 2018-11-05 at 9 48 38 am](https://user-images.githubusercontent.com/895728/48004874-11c1a680-e0e0-11e8-9c32-598e129ad2db.PNG)

serenamarie125 commented:
@rhamilto that is the PF upstream branding, but the diagonal design is used for both upstream and downstream RH products.  From what I remember the bottom right corner is red for downstream and yet gray for upstream .  I'm looking for documentation from Jenny to validate that

spadgett commented:
cc @priley86

priley86 commented:
thanks... going to start trying to port this in now...

rebeccaalpert commented:
This is what I have so far.

When I pulled in the PF styles, I got the PF blue border across the top; let me know if you want that removed.

**Modal example 1:**
<img width="631" alt="screen shot 2018-11-06 at 11 23 18 am" src="https://user-images.githubusercontent.com/7014965/48078277-13aa6900-e1b7-11e8-9839-db92d80b30d7.png">

**Modal example 2:**
<img width="610" alt="screen shot 2018-11-06 at 11 23 44 am" src="https://user-images.githubusercontent.com/7014965/48078282-186f1d00-e1b7-11e8-9a63-6152d142d139.png">

**Navbar without authentication:**
<img width="1252" alt="screen shot 2018-11-06 at 11 23 54 am" src="https://user-images.githubusercontent.com/7014965/48078288-1c9b3a80-e1b7-11e8-96bc-4c9252f41074.png">

**Navbar with authentication:**
Account selected and about on hover:
<img width="1254" alt="screen shot 2018-11-06 at 11 25 19 am" src="https://user-images.githubusercontent.com/7014965/48078293-20c75800-e1b7-11e8-8cd4-a2999d9ce9b4.png">

About selected:
<img width="1253" alt="screen shot 2018-11-06 at 11 25 29 am" src="https://user-images.githubusercontent.com/7014965/48078302-24f37580-e1b7-11e8-87c3-7c60bd4b94ce.png">

No hover/nothing selected:
<img width="1254" alt="screen shot 2018-11-06 at 11 30 20 am" src="https://user-images.githubusercontent.com/7014965/48078417-671cb700-e1b7-11e8-9cfc-cabc82b91553.png">

rebeccaalpert commented:
jenkins rebuild

rebeccaalpert commented:
Addressed feedback.

rebeccaalpert commented:
jenkins rebuild

rebeccaalpert commented:
jenkins rebuild

rebeccaalpert commented:
jenkins rebuild

rebeccaalpert commented:
jenkins rebuild

beanh66 commented:
Thanks @rebeccaalpert! I think these look great. I know you were going to try and make the Red Hat icon a little smaller based on the mockups, but otherwise LGTM. I'm still not sure that downstream background image is right but it's fine until we get something else. 

@serenamarie125 @kybaker let us know if we need a different upstream background image for the OKD version. Thanks!

rebeccaalpert commented:
Made the background and Red Hat logo smaller and added Red Hat to the non-OKD product titles.

Example of non-OKD (product title is dynamic):
<img width="615" alt="screen shot 2018-11-07 at 10 50 48 am" src="https://user-images.githubusercontent.com/7014965/48142659-30a87000-e27b-11e8-9f59-6ccc38504cbf.png">

OKD:
<img width="616" alt="screen shot 2018-11-07 at 10 51 21 am" src="https://user-images.githubusercontent.com/7014965/48142666-343bf700-e27b-11e8-97ce-4dc4a46319ab.png">

rebeccaalpert commented:
Updated the OKD modal.

OKD modal:
<img width="615" alt="screen shot 2018-11-09 at 11 04 02 am" src="https://user-images.githubusercontent.com/7014965/48273561-36cf5580-e40f-11e8-951d-f00d5881541c.png">

alecmerdler commented on frontend/public/components/masthead.tsx:
```diff
@@ -1,30 +1,155 @@
+    this.openDocumentation = this.openDocumentation.bind(this);
+  }
+
+  _checkOpenShiftVersion() {
```
Just use `private` methods instead of `_` prefix.

alecmerdler replied:
Just use `private` methods instead of `_` prefix.

alecmerdler commented on frontend/public/components/masthead.tsx:
```diff
@@ -1,30 +1,155 @@
+    }
+  }
+
+  _checkKubernetesVersion() {
```
Just use `private` methods instead of `_` prefix.

alecmerdler commented on frontend/public/components/masthead.tsx:
```diff
@@ -1,30 +1,155 @@
 
 const developerConsoleURL = (window as any).SERVER_FLAGS.developerConsoleURL;
 
+class HelpMenu extends React.Component<HelpMenuProps, HelpMenuState> {
```
Really should prefer using a functional component and connecting to Redux to fetch the Kubernetes version. It could be useful for other components. Or we could just add it server-side as a JS global to `index.html`.

alecmerdler commented on frontend/public/components/masthead.tsx:
```diff
@@ -1,30 +1,155 @@
+      productTitle = 'OKD';
+  }
+
+  return ({'backgroundImg': backgroundImg, 'logoImg': logoImg, 'logoAlt': logoAlt, 'modalLogo': modalLogoImg, 'modalLogoAlt': modalLogoAlt, 'productTitle': productTitle});
```
Use [object literal shorthand notation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#New_notations_in_ECMAScript_2015) instead of specifying both keys and values.

rebeccaalpert commented:
Dropped OpenShift version for now per discussion with Sam:

<img width="613" alt="screen shot 2018-11-15 at 11 08 47 am" src="https://user-images.githubusercontent.com/7014965/48565565-dc2e7180-e8c6-11e8-89cc-343029df964f.png">

Addressed feedback except for using Redux to fetch the Kubernetes version. We're pushing the account info (which would need Redux) to a follow-on. Would Kubernetes + Redux also be a follow-on?

spadgett commented on frontend/public/components/masthead.tsx:
```diff
@@ -1,30 +1,146 @@
 
+const PFAboutModal = (props) => <AsyncComponent loader={() => import('patternfly-react').then(c => c.AboutModal)} {...props} />;
+const PFAMVersions = (props) => <AsyncComponent loader={() => import('patternfly-react').then(c => c.AboutModal.Versions)} {...props} />;
+const PFAMVersionItem = (props) => <AsyncComponent loader={() => import('patternfly-react').then(c => c.AboutModal.VersionItem)} {...props} />;
```
Does this fire off 3 requests for patternfly-react?

The pattern you want is to put `AboutModal` in a separate module and then load `AboutModal` async. That way you can reference the patternfly-react components directly without async loading from the `AboutModal` component.

rebeccaalpert commented:
Addressed PR feedback.

spadgett commented on frontend/public/components/utils/about-modal.tsx:
```diff
@@ -0,0 +1,60 @@
+}
+
+export type AboutModalProps = {
+  cancel: (e?: Event) => void,
```
`e` shouldn't be optional.

```suggestion
  cancel: (e: Event) => void,
```

https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

spadgett commented on frontend/public/components/utils/about-modal.tsx:
```diff
@@ -0,0 +1,60 @@
+
+export type AboutModalProps = {
+  cancel: (e?: Event) => void,
+  close: (e?: Event) => void,
```
```suggestion
  close: (e: Event) => void,
```

rebeccaalpert commented on frontend/public/components/masthead.tsx:
```diff
@@ -6,25 +6,99 @@ import * as okdLogoImg from '../imgs/okd-logo.svg';
+
 const developerConsoleURL = (window as any).SERVER_FLAGS.developerConsoleURL;
 
+export const BrandingDetails = () => {
```
It's not; it's also used by the logo image in the masthead. (On the left.)

alecmerdler commented on frontend/public/components/utils/about-modal.tsx:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  private checkKubernetesVersion() {
+    k8sVersion()
```
Would prefer to add this to Redux, but we can address that later.

alecmerdler commented on frontend/public/components/utils/about-modal.tsx:
```diff
@@ -0,0 +1,54 @@
+
+  componentDidMount() {
+    this.checkKubernetesVersion();
+    this.setState({ showAboutModal: true });
```
Why are we setting `showAboutModal` to `true` in this method rather than when we initialize `state` above?

rebeccaalpert commented:
I addressed the PR feedback. Still concerned that the async component is introducing a short loading animation before the modal actually shows up.

spadgett commented:
Yeah, I see the small hitch in the animation due to async loading... Let's open a follow on issue for that. We might be able to address as part of #813.

rebeccaalpert commented:
Added [CONSOLE-1041](https://jira.coreos.com/browse/CONSOLE-1041) for the follow-on loading issue and [CONSOLE-1042](https://jira.coreos.com/browse/CONSOLE-1042) for redux.
